### ISSUE TITLE
[css-anchor] Detecting active @position-fallback

### ISSUE THREAD

**Issue by jh3y** [w3c/csswg-drafts#8171]
Currently, we don't have a proposed way to detect which fallback is currently active.

A huge use case for this is if we want to display arrows on elements such as tooltips. If I have a position-fallback that goes from top to bottom:

```
@position-fallback --vertical {
  @try {
    top: anchor(bottom);
  }
  @try {
    bottom: anchor(top);
  }
}
```

If my arrow is styled on the top of my element. But then my element is moved to be on top of the anchor. My arrow now points to space potentially.

Although not possible with the spec as is, being able to use non-inset properties would be interesting inside the fallback. 

```
@position-fallback --vertical {
  @try {
    --on-the-bottom: 1;
    top: anchor(bottom);
  }
  @try {
    --on-the-top: 1;
    bottom: anchor(top);
  }
}
```

You could use that custom property like a boolean flip for the arrow position. But, I imagine there's a more elegant solution to this potentially.

**Comment by xiaochengh** [comment]
There's a fundamental issue: position-fallback is applied at used value time. If we expose the active fallback to CSS, then we allow computed values to depend on used values, which violates the "specified -> computed -> used values" pipeline.

Unless we introduce one more snapshotting: we can snapshot the active fallback at the beginning of a frame and expose it to other properties, like how we currently deal with scrolling.

**Comment by tabatkins** [comment]
The plan is that we'll introduce a new CQ type for this; CQs violate that pipeline in the exact same way, but in a controlled manner. We'll need to be careful that you can't trigger the parent to flip into a different fallback set, tho.

**Comment by bfgeek** [comment]
> The plan is that we'll introduce a new CQ type for this; CQs violate that pipeline in the exact same way, but in a controlled manner. We'll need to be careful that you can't trigger the parent to flip into a different fallback set, tho.

We'll need size containment in both axes for this to work

**Comment by fantasai** [comment]
This isn't fundamentally cyclic, as long as you have a clear and unambiguous way to tie the conditional styles to the position option.

**Comment by kizu** [comment]
I think using something like CQ here is the way, though the tricky thing is the size containment.

The way I think could be the most powerful while avoiding any circularity: an ability to query any container's _closest used fallback_. Critically, not the fallback used on the container itself, as allowing the containment on the children of the positioned element can make things more flexible.

```HTML
<div popover>
  <div class="decorations"></div>
  Some content
</div>
```

With this HTML, for example, we will have a popover which uses some list of fallbacks, and then what we could do is position the `.decorations` with `inset: 0` over the popover, make it a container, and then any pseudo-element (or element) inside this container could query the fallback.

Of course, if the popover itself could be made a container, any children could query the fallback directly, but that will come with more limitations, so an ability to separate the containers from the fallbacks they query could be the most flexible way to go.

**Comment by wesbos** [comment]
I spent some time playing with anchors today and could not find a single example using a tooltip with a "tail" that changes based on it's `position-try` 

For example putting something a tooltip above with `position-area: top center;` 

<img width="497" alt="Screenshot 2024-11-08 at 4 56 38 PM" src="https://github.com/user-attachments/assets/74cd7286-832a-446b-989f-d635deaba8f6">

But then changing the tooltip tail / arrow when it's been flipped with with `position-try-fallbacks: flip-block;` or another `@position-try` 

<img width="496" alt="Screenshot 2024-11-08 at 4 59 04 PM" src="https://github.com/user-attachments/assets/537184f4-e685-4634-8082-b8587ac97cc1">

I was looking for either some sort of keyword for the anchored inline-block side, a style query, or the ability to target more css inside a @position-try past the allowed values. 

All I understand why they don't work, but is this design not possible with CSS anchors?

**Comment by tabatkins** [comment]
At the moment, there is indeed no way to do that. It will have to wait for a container query that can ask what fallback style the abspos is using (and then style a child making the "tail" properly).

**Comment by wesbos** [comment]
OK - thank you for the quick reply!

**Comment by argyleink** [comment]
in addition to having matching tethers/visual indicators like the triangle, it'll also be needed for **directionally aware animation**. a common pattern is to slide down when the select options are under the select button, conversely a slide up animation when the options are above.

**Comment by jpzwarte** [comment]
Is there a way to calculate the offset of a tether if the anchored element is not centered on the anchor for example?

Edit: the above question was rooted in old-fashioned calculating where the tether should be positioned. Somebody pointed out to me that you can also use anchor positioning on the tether!

**Comment by lilles** [comment]
There is an explainer at https://github.com/w3c/csswg-drafts/blob/main/css-anchor-position-1/anchored_container_query.md

TL;DR:

- Introduce a new `container-type:anchored`
- Introduce a new `@container anchored(fallback: <fallback>)` where `<fallback>` matches a value in the comma separated list of `position-try-fallbacks` or `none`.

**Comment by o-t-w** [comment]
The [explainer](https://github.com/w3c/csswg-drafts/blob/main/css-anchor-position-1/anchored_container_query.md) includes “Run different animations based on the position of the anchored element” as an identified need. Does `container-type:anchored` actually solve that use case?

**Comment by lilles** [comment]
> > It uses a new `container-type:anchored` ,and `container anchored(fallback: <integer>)` where `0` means no fallback is used and otherwise a 1-based index into the computed `position-try-fallbacks` value.
> > Disclaimer: I went for the number-indexing for fallbacks for ease of implementation as I wanted to explore implementability, containment requirements, and other limitations.
> 
> As discussed internally _(resharing here for visibility)_: I think we should eventually end up with a syntax where one can refer to a fallback keyword/ident in the query.
> 
> E.g. `@container anchored(fallback: none)`, `@container anchored(fallback: flip-block)`, `@container anchored(fallback: --horizontal)`, etc.
> 
> Referring to numeric indices is error-prone and requires one to adjust the numbers when inserting a new fallback option. But it sure is good enough to start playing with it :)

The prototype in Chrome now uses value-based queries instead of indices as of Canary 139.0.7251.0

**Comment by lilles** [comment]
> The [explainer](https://github.com/w3c/csswg-drafts/blob/main/css-anchor-position-1/anchored_container_query.md) includes “Run different animations based on the position of the anchored element” as an identified need. Does `container-type:anchored` actually solve that use case?

Depends on what you mean. This is e.g. about directional animations like can be seen in the demo: https://codepen.io/lilles/pen/VYLZZqj

If you mean animations between different position-areas, then no.
